### PR TITLE
Resilience

### PR DESCRIPTION
**Pull request by TGfor3**
Added resilience capabilities, including a Retry, Rate Limiter, and a Circuit Breaker

### CODE REVIEW

File: datastore/build.gradle [hunk 1]
```diff
@@ -19,6 +20,12 @@ dependencies {
 	runtimeOnly 'com.h2database:h2'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
 	compile group: 'com.google.code.gson', name: 'gson', version: '2.8.5'
+	implementation'io.github.resilience4j:resilience4j-spring-boot2:1.7.1'
+	implementation 'org.springframework.boot:spring-boot-starter-actuator'
+	implementation 'org.springframework.boot:spring-boot-starter-aop'
+	compile('io.github.resilience4j:resilience4j-spring-boot2:1.7.1')
```
Comment (yspotts) [inline datastore/build.gradle:22]: you shouldnt need these 3 lines....

File: datastore/build.gradle [hunk 2]
```diff
@@ -10,6 +10,7 @@ sourceCompatibility = '11'
 
 repositories {
 	mavenCentral()
+	maven { url 'http://oss.jfrog.org/artifactory/oss-snapshot-local/' }
```
Comment (yspotts) [inline datastore/build.gradle]: are you sure you need this?

File: datastore/src/main/java/com/ds/datastore/BookController.java [hunk 3]
```diff
@@ -63,13 +79,15 @@ private boolean amILeader(){
         return this.storeRepository.findAll().get(0).getServerId().equals(this.leader.getLeader());
     }
 
+    //Retry only worked when placed here, on the more global method but did not work on the Utilities method
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/BookController.java]: should we remove this comment?

File: datastore/src/main/java/com/ds/datastore/BookStoreController.java [hunk 4]
```diff
@@ -40,26 +38,32 @@
...
     }
 
     @PostConstruct
-    private void restartChangedOrNew() throws Exception {
-        this.map.setMap(reclaimMap());
+    public void restartChangedOrNew() throws Exception {
+        try {
+                this.map.setMap(reclaimMap());
+        }catch (Exception e) {
+                System.out.println(e.getClass());
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/BookStoreController.java]: I would prefer this be a log statement

File: datastore/src/main/java/com/ds/datastore/BookStoreController.java [hunk 5]
```diff
@@ -78,12 +81,12 @@ private void registerWithHub() throws Exception {
...
         json.addProperty("id", this.id);
         json.addProperty("address", this.url + "/bookstores/" + this.id);
-        createPutConnection(hubUrl, json);
+        utilities.createConnection(hubUrl, json, this.url, this.id, "PUT");
         logger.info("Server {} connected to network at {}", this.id, this.url);
     }
 
-    private HashMap<Long, String> reclaimMap() throws Exception {
-        HttpResponse<String> response = createGetConnection(hubUrl, this.url, id);
+    public HashMap<Long, String> reclaimMap() throws Exception {
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/BookStoreController.java]: does this still need to be `public`?

File: datastore/src/main/java/com/ds/datastore/Utilities.java [hunk 6]
```diff
@@ -2,61 +2,75 @@
...
+                builder = builder.PUT(HttpRequest.BodyPublishers.ofString(json));
+        }else if (requestType.equals("DELETE")) {
+                builder = builder.DELETE();
+        }
+        HttpRequest request = builder.build();
+        HttpResponse<String> response = HttpClient.newBuilder()
                 .build()
                 .send(request, HttpResponse.BodyHandlers.ofString());
+        if(requestType.equals("POST") && (response.statusCode() != 201 && response.statusCode() != 200) ){
+                System.out.println(response.statusCode());
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/Utilities.java]: lets log something more helpful here and return a more helpful exception (with some content)

File: datastore/src/main/java/com/ds/datastore/Utilities.java [hunk 7]
```diff
@@ -2,61 +2,75 @@
...
-                .headers("Content-Type", "application/json;charset=UTF-8")
-                .PUT(HttpRequest.BodyPublishers.ofString(json))
-                .build();
-        return HttpClient.newBuilder()
-                .build()
-                .send(request, HttpResponse.BodyHandlers.ofString());
+    @Retry(name = "retry")
+    @CircuitBreaker(name = "#root.args[0]", fallbackMethod = "fallback")
+    public Optional<HttpResponse<String>> createConnectionCircuitBreaker(String address, JsonObject jso, String serverAddress, Long id, String requestType) throws Exception{
+        System.out.println("utilities called");
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/Utilities.java]: lets use log statements

File: datastore/src/main/java/com/ds/datastore/Utilities.java [hunk 8]
```diff
@@ -2,61 +2,75 @@
...
-        HttpRequest request = HttpRequest.newBuilder()
-                .uri(new URI(address))
-                .headers("Content-Type", "application/json;charset=UTF-8")
-                .DELETE()
-                .build();
-        HttpClient.newBuilder()
-                .build()
-                .send(request, HttpResponse.BodyHandlers.ofString());
+    private Optional<HttpResponse<String>> fallback(String address, JsonObject jso, String serverAddress, Long id, String requestType, RuntimeException e) {
+        System.out.println("falling back");
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/Utilities.java]: log statement and lets remove the commented code

File: hub/src/main/java/com/hub/HubController.java [hunk 9]
```diff
@@ -95,20 +94,16 @@ protected Long addServer(@RequestBody String json) throws Exception {
         return serverId;
     }
 
-    private void sendLeader() throws IOException {
+    private void sendLeader() throws Exception {
```
Comment (yspotts) [inline hub/src/main/java/com/hub/HubController.java:97]: sending the leader might benefit from a retry

File: datastore/build.gradle (hunk 1 above)
Comment (TGfor3) [reply datastore/build.gradle:22]: Assuming this isnt a joke, VSCode wouldn't work without them

Comment (yspotts) [reply datastore/build.gradle:22]: im not joking at all. What if you took out the 3 `implementation` lines then?

File: datastore/src/main/java/com/ds/datastore/Utilities.java [hunk 10]
```diff
@@ -44,7 +46,7 @@
                 .build()
                 .send(request, HttpResponse.BodyHandlers.ofString());
         if(requestType.equals("POST") && (response.statusCode() != 201 && response.statusCode() != 200) ){
-                System.out.println(response.statusCode());
+                logger.warn("Did not receive a successful status code");
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/Utilities.java]: would probably be helpful to include what status code was received.

File: datastore/src/main/java/com/ds/datastore/Utilities.java [hunk 11]
```diff
@@ -44,7 +46,7 @@
...
                 throw new RuntimeException();
```
Comment (yspotts) [inline datastore/src/main/java/com/ds/datastore/Utilities.java:50]: and I still want a more helpful message here :)

File: hub/src/main/java/com/hub/HubController.java (hunk 9 above)
Comment (SethJacobs) [reply hub/src/main/java/com/hub/HubController.java:97]: Anytime a server starts back up it calls getLeader so the hub shouldn't have to retry sending the leader

Comment (yspotts) [reply hub/src/main/java/com/hub/HubController.java:97]: good point

File: datastore/src/main/java/com/ds/datastore/Utilities.java (hunk 11 above)
Comment (TGfor3) [reply datastore/src/main/java/com/ds/datastore/Utilities.java:50]: Changed it to : {} received, POST failed", response.statusCode()

Comment (yspotts) [reply datastore/src/main/java/com/ds/datastore/Utilities.java:50]: I meant in the `RuntimeException`

Comment (TGfor3) [reply datastore/src/main/java/com/ds/datastore/Utilities.java:50]: We only threw that to force a retry. Didn't really intend it to be descriptive of what actually happened

Comment (yspotts) [reply datastore/src/main/java/com/ds/datastore/Utilities.java:50]: oh I see.

**Review by yspotts — APPROVED** [review]

**Pull request merged**